### DISCUSSION TITLE
Modifying copy button behaviour in code blocks

### DISCUSSION THREAD

**Discussion by mfridman** [facebook/docusaurus#5227] [Q&A] [answered]
Is it possible to have a code block such as:

```bash
$ echo "hello world" | cat 
```

And have the copy button not include the prefixed `$` ?

Quite often you may want to specify a terminal command, denoted with the `$` identifier .. but not actually have it copied along with the command.

**Comment by slorber** [comment, marked as answer]
I don't think it's possible currently. 
We also had a similar issue about diff code-blocks:
https://github.com/facebook/docusaurus/issues/4698

Was wondering if adding a `$` prefix shouldn't be a Prism highlighting theme concern that should be left of your actual code block? 
Maybe worth looking in Prism issues to see if this ever came up

**Reply by mfridman** [reply]
Unfortunately couldn't find anything of use in Prism, and it's probably less a Prism concern as you mentioned.

Could you elaborate on this?

> Was wondering if adding a $ prefix shouldn't be a Prism highlighting theme concern that should be left of your actual code block

---

I presume this is a common enough annoyance that might warrant a generic solution? 

Maybe a user-supplied list of tokens, characters or even regex that could be ignored during copying, and substituted with an equivalent amount of space?

So in our case we could supply the tokens: `$`, `-`, `+` and during copying if the line begins with any of these tokens it gets substituted with a space equivalent to the number of characters (in this example just 1 space).

**Reply by mfridman** [reply]
Hmm, this plugin could accomplish what we'd need:

https://prismjs.com/plugins/command-line

But I'm not sure if such plugins are supported through docusaurus.

---

Update: answer is no, cannot add plugins due to upstream project `prism-react-renderer` https://github.com/facebook/docusaurus/discussions/4829#discussion-3376093

So, any way we can add basic support for this common use case in docusaurus @slorber ?

**Reply by slorber** [reply]
@mfridman I'm not against but this is low-priority for me, so waiting for a community contribution.

In your case, the $ does not have to be part of the code block (unlike the diff case). 

You can just add the $ prefix from the outside.
Look at how this example is adding line numbers:
https://codesandbox.io/s/prism-react-renderer-example-u6vhk?file=/src/WithLineNumbers.js

You can swizzle our existing CodeBlock comp and adapt it to your own needs, also override the copy/pasting logic

In your position, I'd just use some custom CSS:

```css
div.bash span.token-line::before {
    content: '$';
    padding: 0 5px;
    color: red;
}
```

![image](https://user-images.githubusercontent.com/749374/127335787-b8f0fa31-bed2-4701-9ea7-26bdb0389c38.png)

**Reply by mfridman** [reply]
Thank you. The following CSS should do the trick, assuming a codeblock specifying the `terminal` language

```
div.terminal span.plain:first-child::before {
  content: '$';
  padding: 0 5px;
  color: red;
}
```

We also wanted to solve the `diff` issue similar to https://github.com/facebook/docusaurus/issues/4698.

So I'll take a look at this comment

> You can swizzle our existing CodeBlock comp and adapt it to your own needs, also override the copy/pasting logic

Again, thank you for the info.

**Reply by slorber** [reply]
Great

Swizzle is good for testing but ultimately if your own solution works for your site we'd be happy to fix this for all on Docusaurus core, so don't hesitate to send a PR :)

I don't think removing +/- is a good enough solution, maybe it can mess up with some languages in weird ways? I'd prefer to have this sandboxed to a few "languages" like "diff" only

**Reply by mfridman** [reply]
Agreed this would have to be language-specific _and_ backwards compatible in an explicit "opt-in"-kind of way as to not break anyone who wouldn't want this behaviour.

I'll mark this GitHub Discussion completed for now, as this likely belongs in a larger issue and/or PR discussion for various community input and visibility.

**Comment by matkoch** [comment]
Just for others that are interested in this. I have a slightly different approach using "magic comments" of prism. In `docusaurus.config.js`, I put the following:

```
      prism: {
        theme: lightCodeTheme,
        darkTheme: darkCodeTheme,
        additionalLanguages: ['powershell', 'csharp', 'yaml'],
        magicComments: [
          // Remember to extend the default highlight class name as well!
          {
            className: 'theme-code-block-highlighted-line',
            line: 'highlight-next-line',
            block: {start: 'highlight-start', end: 'highlight-end'},
          },
          {
            className: 'code-block-terminal-command',
            line: 'terminal-command',
          },
        ],
      },
```

My `custom.css` got this portion:

```
.code-block-terminal-command:before {
    content: '$';
    margin-right: 10px;
}
```

And in my code blocks I do the following:

````
```powershell
# terminal-command
dotnet watch run -- [arguments..] --continue
... output of command
```
````

This has the benefit that the `$` is not shown for every line and I can still show the output of the command in the same code block. For instance:

```
$ dotnet watch run -- [arguments..] --continue
... output of command
```

To be fair, this doesn't help with the copy button, but I still found it a nicer approach for some use-cases :)
